Support closure variables in nested functiuon defs

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170705

Changelogs:
- Need to update some vmap tests because as we rebuild nested function, we need to realize the values from closure which affects graph input ordering 
- Need to add some state to nested function to break cycles 
- When closure refers to nested function itself, break cycle gracefully 
- Subclassed NotImplementedError to catch exception in export better. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan committed:
Support closure variables in nested functiuon defs

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support closure variables in nested functiuon defs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 reviewed (commented):
Can you add more tests to test/dynamo? Also make sure to test more complicated closure objects like lists/dicts, objects, etc. Check that a closure object that doesn't have as_python_constant() defined errors out gracefully.

tugsbayasgalan committed:
Update on "Support closure variables in nested functiuon defs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support closure variables in nested functiuon defs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support closure variables in nested functiuon defs"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/variables/base.py:
```diff
@@ -218,6 +218,11 @@ def __init__(self, vt: "VariableTracker") -> None:
 
 
+# Raised when get_function() cannot convert a nested function to a Python function.
+class ClosureConversionError(NotImplementedError):
```
Move to functions.py since it's mainly relevant to that file?

williamwen42 commented on torch/_dynamo/output_graph.py:
```diff
@@ -1695,22 +1695,19 @@ def compile_subgraph(
-                            raise
+                    except ClosureConversionError as e:
+                        unimplemented(
+                            gb_type="nested function with closure in output",
```
```suggestion
                            gb_type="nested function with non-constructible closure in output",
```

williamwen42 commented on torch/_dynamo/output_graph.py:
```diff
@@ -1695,22 +1695,19 @@ def compile_subgraph(
+                            context="as_python_constant for out_spec",
+                            explanation=(
+                                "Cannot return a nested function with closure from a compiled function. "
+                                "Dynamo cannot reconstruct the function."
```
```suggestion
                                "Dynamo failed to construct the function defined in the compiled region with closure objects."
```

williamwen42 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1590,12 +1599,56 @@ def python_type(self) -> type:
     def get_function(self) -> types.FunctionType:
+        from .base import AsPythonConstantNotImplementedError, ClosureConversionError
+
+        if self._converting:
```
nit: could we just use a set (of id()) instead of adding a new field?

tugsbayasgalan replied:
Done

williamwen42 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1590,12 +1599,56 @@ def python_type(self) -> type:
+        try:
+            return self._get_function_impl()
+        except AsPythonConstantNotImplementedError as e:
+            raise ClosureConversionError(str(e)) from e
```
```suggestion
            raise ClosureConversionError("failed to convert closure cell to Python constant") from e
```

tugsbayasgalan committed:
Update on "Support closure variables in nested functiuon defs"


Changelogs:
- Need to update some vmap tests because as we rebuild nested function, we need to realize the values from closure which affects graph input ordering 
- Need to add some state to nested function to break cycles 
- When closure refers to nested function itself, break cycle gracefully 
- Subclassed NotImplementedError to catch exception in export better. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 commented on test/dynamo/test_misc.py:
```diff
@@ -10501,6 +10501,102 @@ def start():
+            def inner():
+                return x * config["scale"] + sum(items) + config["nested"]["offset"]
+
+            return inner()
```
Can you confirm if this test runs into the NestedFunction NotImplemented before your changes?

williamwen42 commented on torch/_dynamo/graph_break_registry.json:
```diff
@@ -980,6 +980,17 @@
       "Hints": []
     }
   ],
+  "GB2692": [
```
This should not be a new graph break - we should append the new entry to the existing gb registry entry

tugsbayasgalan committed:
Update on "Support closure variables in nested functiuon defs"


Changelogs:
- Need to update some vmap tests because as we rebuild nested function, we need to realize the values from closure which affects graph input ordering 
- Need to add some state to nested function to break cycles 
- When closure refers to nested function itself, break cycle gracefully 
- Subclassed NotImplementedError to catch exception in export better. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support closure variables in nested functiuon defs"


Changelogs:
- Need to update some vmap tests because as we rebuild nested function, we need to realize the values from closure which affects graph input ordering 
- Need to add some state to nested function to break cycles 
- When closure refers to nested function itself, break cycle gracefully 
- Subclassed NotImplementedError to catch exception in export better. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support closure variables in nested functiuon defs"


Changelogs:
- Need to update some vmap tests because as we rebuild nested function, we need to realize the values from closure which affects graph input ordering 
- Need to add some state to nested function to break cycles 
- When closure refers to nested function itself, break cycle gracefully 
- Subclassed NotImplementedError to catch exception in export better. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

williamwen42 reviewed (approved):
Is it possible to add a test for nested function get_function w/ closure (i.e. export blockmask test) that doesn't require CUDA?
